### PR TITLE
Fix some seq retention issues

### PR DESCRIPTION
**Pull request by cichli**
I first discovered that

``` clojure
(binding [*print-length* 10]
  (fipp.edn/pprint (range)))
```

would never return, because we were comparing `print-length` to `(count xs)` - which is no good on an infinite sequence. Switched to using `(seq? (drop print-length xs))`, and added extra tests to ensure a) this works for infinite sequences and b) a potential off-by-one error has not been made.

Further poking around showed that the following would cause heap exhaustion:

``` clojure
(binding [*out* (proxy [java.io.Writer] []
                  (write [_])
                  (flush [])
                  (close []))]
  (fipp.edn/pprint (range 100000000)))
```

There were two places causing memory to leak:
- `serialize` used `mapcat`, which causes full realisation of chunked sequences (I think this is a bug in the `[xs ys & zs]` arity of `concat`; it seems `zs` is being rooted inside the `cat` closure. Not sure if there's a JIRA related to this but I'll take a look). Wrap `doc` in a call to `unchunk` to ensure the chunked sequence path is not taken in `concat`.
- `Eduction`s hold a reference to their `coll` field - switch to using `sequence`, which does not hold onto anything when used with `run!`.

### CODE REVIEW

**Comment by brandonbloom** [comment]
Solid detective work. A few questions/comments coming inline...

File: src/fipp/engine.cljc [hunk 1]
```diff
@@ -5,6 +5,12 @@
             [fipp.deque :as deque]))
 
 
+(defn ^{:private true} unchunk [s]
```
Comment (brandonbloom) [inline src/fipp/engine.cljc]: I'd suggest the `^:private` shorthand notation, but in general _everything_ in this file is private except `pprint-document`. I'd rather just document that fact in the readme than give the illusion that the other stuff is also public by hinting on one function as private.

File: src/fipp/engine.cljc [hunk 2]
```diff
@@ -229,10 +235,9 @@
 (defn pprint-document [document options]
   (let [options (merge {:width 70} options)]
     (->> (serialize document)
-         (eduction
-           annotate-rights
-           (annotate-begins options)
-           (format-nodes options))
+         (sequence (comp annotate-rights
```
Comment (brandonbloom) [inline src/fipp/engine.cljc]: Does this significantly hurt performance by forcing a lazy sequence to be realized with all of the intermediate data structures that entails?

File: src/fipp/engine.cljc (hunk 1 above)
Comment (brandonbloom) [inline src/fipp/engine.cljc]: There's also this ticket: http://dev.clojure.org/jira/browse/CLJ-1218 - Is this faster maybe?

File: src/fipp/engine.cljc (hunk 2 above)
Comment (cichli) [inline src/fipp/engine.cljc]: There is an overhead to creating a chunked sequence when using `sequence`, yes, and naively I thought this would be faster:

```clojure
(->> (serialize document)
     (transduce (comp annotate-rights
                      (annotate-begins options)
                      (format-nodes options))
                (completing #(print %2))
                nil))
```

This is essentially what `(run! the-eduction print)` was doing, without the indirection of an eduction.

But in my testing, the `sequence` version is ~10% faster (using the existing benchmark and a silly microbenchmark printing `(range 100000)`). I _think_ this is because the [`IChunkedSeq`](https://github.com/clojure/clojure/blob/bc186508ab98514780efbbddb002bf6fd2938aee/src/clj/clojure/core/protocols.clj#L131) path of `InternalReduce` is being used rather than the [default](https://github.com/clojure/clojure/blob/bc186508ab98514780efbbddb002bf6fd2938aee/src/clj/clojure/core/protocols.clj#L158). I presume the chunking optimisation is winning over the overhead of having to create the chunked sequence.

File: src/fipp/engine.cljc (hunk 1 above)
Comment (cichli) [inline src/fipp/engine.cljc]: r.e. `{:private true}` - I'd personally just use `defn-` (or like you say, not worry about making it private), but I know that and the short form of metadata can rankle some people so played it safe :-). Will just go ahead and make it public.

r.e. the mapcat issue - I think the bug here is actually a little bit more subtle. I had seen that issue, and tried using the version of `join` that @ghadishayban includes in the comments there, but still exhausted the heap. I've narrowed it down to this:

```clojure
;; doesn't exhaust heap
(dorun (concat (range) (range)))
(dorun (concat (range) (range) (range)))
(dorun (apply concat [(range) (range)]))

;; exhausts heap
(dorun (apply concat [(range) (range) (range]))

;; doesn't exhaust heap
(dorun (apply concat (unchunk [(range) (range) (range)])))
```

Turns out it's because of sequence chunking - when the `zs` argument to concat is chunked (as it is when using `apply`), its entire first chunk will be retained in memory (and hence the fully realised range contained in that chunk will be retained too). Thanks to @bronsa for pointing this out to me.

```clojure
user=> (defn x [x & y] (println (class y)))
#'user/x
user=> (x 1 2 3 4)
clojure.lang.ArraySeq
nil
user=> (apply x [1 2 3 4])
clojure.lang.PersistentVector$ChunkedSeq
nil
user=> (apply x (unchunk [1 2 3 4]))
clojure.lang.Cons
```

File: src/fipp/engine.cljc (hunk 2 above)
Comment (brandonbloom) [inline src/fipp/engine.cljc]: OK, I'll benchmark this a bit myself when I get some spare time.

**Comment by brandonbloom** [comment]
Maybe I'm missing something, but I'm seeing heap exhaustion for that even with both of these commits applied. Are you sure that applying these make that `(range 100000000)` test case work?

Thinking about this a bit more: Is there even a use case for pretty printing extremely large unretained lazy seqs? It seems like most pretty-printing needs would proceed from some root value that will retain the entire tree of values being printed in the calling code anyway. Do you have a genuine need for these changes?

**Comment by cichli** [comment]
I'm definitely not seeing heap exhaustion for that example, even if I change the object being printed to something that you'd expect to root the sequence like `{:a {:b {:c (range 100000000)}}}` or `(atom (range 100000000))`. The same applies if I lift the object being printed into a local binding with `let` - isn't this just normal Clojure locals clearing ensuring that nothing retains onto the object if it doesn't need to?

The way I'm testing this is attaching to the process with YourKit and watching the heap usage graph. Forcing a manual GC will always take the heap usage back to the same baseline.

I asked a colleague (@bronsa) to check and he's seeing the same results as me. We're both using Java 8 FWIW - are you using Java 7? There were some escape analysis improvements in HotSpot 8 that may be responsible for this.

> Thinking about this a bit more: Is there even a use case for pretty printing extremely large unretained lazy seqs?

I'd say that being able to print potentially larger-than-memory lazy data structures is definitely useful, for example:

```clojure
(require '[clojure.java.io :as io])
(require '[fipp.edn :as edn])

(binding [*out* (io/writer (io/file "/tmp/foo.edn"))]
  (edn/pprint (... some DB query that returns a larger-than-memory lazy seq ...)))
```

Granted, in this case it would be possible to `(run! edn/pprint xs)`, but that only solves the case where the root thing I'm printing is an unrealised lazy sequence - but not if it's e.g. a map that contains an unrealised lazy sequence as one of its values (which by my testing above, will not be retained in memory).

**Comment by brandonbloom** [comment]
Thanks for the additional details. Given the testing complexity and unproven impact here, I probably won't really dig in to this for a while, just in the interest of my own time availability. If memory exhaustion becomes a regular problem in practice, please let me know and I'll find some time to tool around with YourKit.

**Comment by brandonbloom** [comment]
Sorry I forgot about this PR. The main fix for the `(count infinite-seq)` issue got incorporated by f0c6ab05 a while ago.

I couldn't reproduce the heap exhaustion issues with up-to-date Clojure and JVM. Please shout if this is still a problem for you in a production context.
